integrations/go/workload-executor.go

```diff
@@ -0,0 +1,309 @@
+		key := elem.Key()
+		val := elem.Value()
+
+		switch key {
+		case "filter":
+			filter = val.Document()
+		case "update":
+			update, err = createUpdate(val)
+			if err != nil {
+				return nil, err
+			}
+		default:
+			str := fmt.Sprintf("unrecognized updateOne option: %v", key)
+			panic(str)
+		}
+	}
+	if opts.Upsert == nil {
+		opts = opts.SetUpsert(false)
+	}
+
+	return coll.UpdateOne(context.Background(), filter, update, opts)
+}
+
+func verifyInsertOneResult(actualResult *mongo.InsertOneResult, expectedResult interface{}) bool {
+	if expectedResult == nil {
+		return true
+	}
+
+	var expected mongo.InsertOneResult
+	if bson.Unmarshal(expectedResult.(bson.Raw), &expected) != nil {
+		return false
+	}
+
+	expectedID := expected.InsertedID
+	if f, ok := expectedID.(float64); ok && f == math.Floor(f) {
+		expectedID = int32(f)
+	}
+
+	return expectedID == nil || (actualResult != nil && expectedID == actualResult.InsertedID)
+}
+
+func verifyCursorResult(cur *mongo.Cursor, result interface{}) bool {
+	if result == nil {
+		return true
+	}
+
+	if cur == nil {
+		return false
+	}
+
+	defer func() {
+		err := cur.Close(context.Background())
+		if err != nil {
+			panic(err)
+		}
+	}()
+
+	for _, expected := range result.(bson.A) {
+		if !cur.Next(context.Background()) {
+			return false
+		}
+		if !bytes.Equal(expected.(bson.Raw), cur.Current) {
+			return false
+		}
+	}
+
+	if cur.Next(context.Background()) {
+		return false
+	}
+	return cur.Err() == nil
+}
+
+func verifyUpdateResult(res *mongo.UpdateResult, result interface{}) bool {
+	if result == nil {
+		return true
+	}
+
+	var expected struct {
+		MatchedCount  int64 `bson:"matchedCount"`
+		ModifiedCount int64 `bson:"modifiedCount"`
+		UpsertedCount int64 `bson:"upsertedCount"`
+	}
+	err := bson.Unmarshal(result.(bson.Raw), &expected)
+	if err != nil {
+		return false
+	}
+
+	if expected.MatchedCount != res.MatchedCount {
+		return false
+	}
+	if expected.ModifiedCount != res.ModifiedCount {
+		return false
+	}
+
+	actualUpsertedCount := int64(0)
+	if res.UpsertedID != nil {
+		actualUpsertedCount = 1
+	}
+	return expected.UpsertedCount == actualUpsertedCount
+}
+
+func executeCollectionOperation(coll *mongo.Collection, op *operation) (bool, error) {
+	switch op.Name {
+	case "insertOne":
+		res, err := executeInsertOne(coll, op.Arguments)
+		return verifyInsertOneResult(res, op.Result), err
+	case "find":
+		cursor, err := executeFind(coll, op.Arguments)
+		return verifyCursorResult(cursor, op.Result), err
+	case "updateOne":
+		res, err := executeUpdateOne(coll, op.Arguments)
+		return verifyUpdateResult(res, op.Result), err
+	}
+	return false, errors.New("unrecognized collection operation: " + op.Name)
+}
+
+func runOperation(coll *mongo.Collection, op *operation) (bool, error) {
+	// execute the command on the given object
+	if op.Object == "collection" {
+		return executeCollectionOperation(coll, op)
+	}
+	return false, errors.New("unrecognized object: " + op.Name)
```

Comment:
According to https://mongodb-labs.github.io/drivers-atlas-testing/spec-test-format.html#test-scenario-format-specification object could also be "database". Though I assume if this is passing, perhaps that is not the case.

I don't see any mention of "database" in the operations of https://github.com/mongodb-labs/drivers-atlas-testing/tree/master/tests. But perhaps this should panic instead of returning an error? A hard crash may be easier to interpret if "database" was added to the format later. Currently I think this error message won't be printed anywhere.

Reply:
The idea behind allowing object to be "database" was that we _could_ write tests that ran operations on Database objects but none of the existing planned maintenance tests do this. Eventually, much of the format used for these tests will match what comes out of the Unified Test Format project so I don't see any reason to add support for database at this point in time.